### PR TITLE
implementation of deepfool

### PR DESCRIPTION
**Pull request by cihangxie**
implementation of deepfool algorithm https://arxiv.org/pdf/1511.04599.pdf

### CODE REVIEW

**Comment by cihangxie** [comment]
we have signed it. Sorry for the inconvenience.

**Comment by goodfeli** [comment]
You need to say exactly "I signed it!" to or googlebot won't understand

**Comment by cihangxie** [comment]
I signed it!

File: cleverhans_tutorials/mnist_tutorial_deepfool.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
+# adapted from mnist_tutorial_cw.py
```
Comment (npapernot) [inline cleverhans_tutorials/mnist_tutorial_deepfool.py]: @goodfeli : Do you think we should keep reproducing tutorials for each attack? This tutorial is adapted from the CW tutorial, which is itself adapted from the JSMA tutorial. 

Perhaps it would make more sense to write one tutorial and have a flag used to choose the attack algorithm?

File: cleverhans/attacks.py [hunk 2]
```diff
@@ -749,6 +749,79 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
         self.clip_max = clip_max
 
 
+class DeepFool(Attack):
+
+    """
+    DeepFool paper link: "https://arxiv.org/pdf/1511.04599.pdf"
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could you add a short description of the attack to explain what the attack is doing in a few words?

File: cleverhans/attacks.py [hunk 3]
```diff
@@ -749,6 +749,79 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
         self.clip_max = clip_max
 
 
+class DeepFool(Attack):
```
Comment (npapernot) [inline cleverhans/attacks.py]: There is no test for the new class. Could you add them to the `tests_tf/test_attacks.py` file?

File: cleverhans/attacks.py [hunk 4]
```diff
@@ -749,6 +749,79 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
...
+            return deepfool_batch(self.sess, x, preds, logits, grads, x_val,
+                                  self.nb_candidate, self.overshoot,
+                                  self.max_iter, self.clip_min, self.clip_max,
+                                  self.nb_classes)
+        wrap = tf.py_func(deepfool_wrap, [x], tf.float32)
+
+        return wrap
+
+    def parse_params(self, nb_candidate=10, overshoot=0.02, max_iter=50,
+                     nb_classes=1001, clip_min=0., clip_max=1., **kwargs):
```
Comment (npapernot) [inline cleverhans/attacks.py]: The docstring is missing here. It should provide a description of the arguments. (It's OK to copy it over from `attacks_tf.py`

File: cleverhans/attacks.py [hunk 5]
```diff
@@ -749,6 +749,79 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
...
+                                  'nb_candidate', 'nb_classes']
+
+        if not isinstance(self.model, Model):
+            self.model = CallableModelWrapper(self.model, 'logits')
+
+    def generate(self, x, **kwargs):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        :param x: The model's symbolic inputs.
+        :param nb_candidate: The number of adversarial class candidates
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could you elaborate on what is the number of adversarial class candidates?

File: cleverhans/attacks.py [hunk 6]
```diff
@@ -749,6 +749,79 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
...
+
+        # Define graph
+        def deepfool_wrap(x_val):
+            return deepfool_batch(self.sess, x, preds, logits, grads, x_val,
+                                  self.nb_candidate, self.overshoot,
+                                  self.max_iter, self.clip_min, self.clip_max,
+                                  self.nb_classes)
+        wrap = tf.py_func(deepfool_wrap, [x], tf.float32)
+
+        return wrap
```
Comment (npapernot) [inline cleverhans/attacks.py]: It's OK to just write one line instead of defining wrap = and then returning wrap. (It should fit under 80 characters)

File: cleverhans/attacks_tf.py [hunk 7]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    :param nb_candidate: The number of adversarial class candidates
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for deepfool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :param nb_classes: Number of model output classes
+    :return: Adversarial examples
+    """
+
+    X_adv = np.zeros(X.shape)
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Why is this line necessary?

File: cleverhans/attacks_tf.py [hunk 8]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    X_adv = deepfool_attack(sess, x, pred, logits, grads, X, nb_candidate,
+                            overshoot, max_iter, clip_min, clip_max, feed=feed)
+
+    return np.asarray(X_adv, dtype=np.float32)
+
+
+def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
+                    overshoot, max_iter, clip_min, clip_max, feed=None):
+    """
+    TensorFlow implementation of the deepfool.
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: You can remove the word "the" in this sentence. (and perhaps capitalize deepfool if the paper writes it DeepFool as well?)

File: cleverhans/attacks_tf.py [hunk 9]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    Paper link: see https://arxiv.org/pdf/1511.04599.pdf
+    :param sess: TF session
+    :param x: The input placeholder
+    :param predictions: The model's symbolic output (linear output,
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Can you add the word logits in this description somewhere? It is a standard way to refer to the pre-softmax activations.

Comment (npapernot) [reply cleverhans/attacks_tf.py]: Why is this argument needed if the argument logits (documented at line 767) is here as well?

File: cleverhans/attacks_tf.py [hunk 10]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+        pre-softmax)
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
+    :param sample: Numpy array with sample input
+    :param nb_candidate: The number of adversarial class candidates
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for deepfool
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Capitalize DeepFool if needed

File: cleverhans/attacks_tf.py [hunk 11]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    """
+    TensorFlow implementation of the deepfool.
+    Paper link: see https://arxiv.org/pdf/1511.04599.pdf
+    :param sess: TF session
+    :param x: The input placeholder
+    :param predictions: The model's symbolic output (linear output,
+        pre-softmax)
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Try to define more precisely which gradients are needed here. It is good also to say that people should use `gradient_graph` to produce these gradients.

File: cleverhans/attacks_tf.py [hunk 12]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    """
+    import copy
+
+    adv_x = copy.copy(sample)
+    # Initialize the loop variables
+    iteration = 0
+    current = utils_tf.model_argmax(sess, x, logits, adv_x, feed=feed)
+    if current.shape == ():
+        current = np.array([current])
+    w = np.squeeze(np.zeros(sample.shape[1:4]))  # same shape as original image
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Does this imply that `x` should be an image and have the usual TF dimension ordering? If yes, then could you add a check at the beginning of the function that throws an error if the input `x` does not have a compatible shape?

File: cleverhans/attacks_tf.py [hunk 13]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+                  format(max_iter))
+    # Repeat this main loop until we have achieved misclassification
+    while (np.any(current == original) and iteration < max_iter):
+
+        if iteration % 5 == 0 and iteration > 0:
+            _logger.info("Attack result at iteration {} is {}".format(
+                iteration,
+                current))
+        gradients = sess.run(grads, feed_dict={x: adv_x})
+        f = sess.run(predictions, feed_dict={x: adv_x})
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Could you use a more descriptive variable name?

File: cleverhans/attacks_tf.py [hunk 14]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+        for idx in range(sample.shape[0]):
+            pert = np.inf
+            if current[idx] != original[idx]:
+                continue
+            for k in range(1, nb_candidate):
+                w_k = gradients[idx, k, ...] - gradients[idx, 0, ...]
+                f_k = f[idx, k] - f[idx, 0]
+                pert_k = (abs(f_k) + 0.00001)/np.linalg.norm(w_k.flatten())
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Can you add spaces around the `/` operator?

Comment (npapernot) [reply cleverhans/attacks_tf.py]: Is this the exact algorithm used in the paper? If no, where is the value `0.00001` coming from?

File: cleverhans/attacks_tf.py [hunk 15]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    _logger.info("{} out of {}".format(sum(current != original),
+                                       sample.shape[0]) +
+                 " becomes adversarial examples at iteration {}".format(
+                     iteration))
+    # need to clip this image into the given range
+    adv_x = np.clip((1+overshoot)*r_tot + sample, clip_min, clip_max)
+    return adv_x
+
+
+def gradient_graph(predictions, x, nb_candidate):
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Can you add tests for this function to `tests_tf/test_attacks_tf.py`?

File: cleverhans/attacks_tf.py [hunk 16]
```diff
@@ -726,3 +726,126 @@ def compare(x, y):
...
+    """
+
+    X_adv = np.zeros(X.shape)
+    X_adv = deepfool_attack(sess, x, pred, logits, grads, X, nb_candidate,
+                            overshoot, max_iter, clip_min, clip_max, feed=feed)
+
+    return np.asarray(X_adv, dtype=np.float32)
+
+
+def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Can you add tests for this function to `tests_tf/test_attacks_tf.py`?

**Comment by goodfeli** [comment]
I would vote no more tutorials, no new flags running different attacks in the existing tutorials.

A whole tutorial demonstrating an attack is more work to maintain than the underlying library functionality of the attack itself.

If we add more flags to the existing tutorials to make them implement more features, they become too complicated for the purpose of being an easy-to-read document for people to learn from.

If we really want an all purpose script that can train any model or run any attack, that's not really a tutorial anymore, and we should make an official trainer script or attack generation script.

**Comment by bairdzhang** [comment]
I signed it!

**Comment by bairdzhang** [comment]
I've signed the CLA, but it says 'Failure: CLAs are signed, but unable to verify author consent', could you please tell me what am I missing? And there are few commits I pushed with email address 'bairdzhang@users.noreply.github.com', and how can I sign CLA for those commits? Thanks!

**Comment by goodfeli** [comment]
I think the easiest thing might be:

1) Close this PR

2) Do a rebase and "amend" the commits using "bairdzhang@users.noreply.github.com" to use the right e-mail address

3) Open a new PR

Sorry for the trouble with Googlebot

**Comment by npapernot** [comment]
Just for the record, I agree with Ian's comment above: we should limit the number of tutorials. They don't bring much value now that the Attacks interface standardized the API across different algorithms.

File: cleverhans/attacks_tf.py (hunk 16 above)
Comment (cihangxie) [reply cleverhans/attacks_tf.py]: We have added DeepFool test to /tests_tf/test_attacks.py, is it still necessary to create a test for this function?

Comment (npapernot) [reply cleverhans/attacks_tf.py]: It should be fine, thanks
